src/lib.rs

```diff
@@ -1265,36 +1260,43 @@ impl Build {
             wait_on_child(&cmd, &program, &mut child.0)?;
         }
 
-        // Reacquire our process's token before we proceed, which we released
-        // before entering the loop above.
-        if reacquire {
-            server.acquire_raw()?;
-        }
-
         return Ok(());
 
         /// Returns a suitable `jobserver::Client` used to coordinate
         /// parallelism between build scripts.
         fn jobserver() -> &'static jobserver::Client {
+            use std::mem::MaybeUninit;
+
             static INIT: Once = Once::new();
-            static mut JOBSERVER: Option<jobserver::Client> = None;
+            static mut JOBSERVER: MaybeUninit<jobserver::Client> = MaybeUninit::uninit();
 
             fn _assert_sync<T: Sync>() {}
             _assert_sync::<jobserver::Client>();
 
             unsafe {
                 INIT.call_once(|| {
-                    let server = default_jobserver();
-                    JOBSERVER = Some(server);
+                    JOBSERVER.write(default_jobserver());
                 });
-                JOBSERVER.as_ref().unwrap()
+                JOBSERVER.assume_init_ref()
             }
         }
 
         unsafe fn default_jobserver() -> jobserver::Client {
             // Try to use the environmental jobserver which Cargo typically
             // initializes for us...
             if let Some(client) = jobserver::Client::from_env() {
+                match client.available() {
+                    Ok(0) | Err(_) => {
+                        // There is no available tokens.
+                        // This is probably because `cc::Build::compile_objects` is called in
+                        // `build.rs` where the `cargo` holds the token.
+                        //
+                        // In that case, we will release one raw token just to prevent deadlock,
+                        // but this will increase the allowed global concurrency by one.
+                        let _ = client.release_raw();
+                    }
```

Comment:
@joshtriplett I implemented what you requested (only acquire and release token when actually compiling) but with a catch.

`cargo` would acquire one token before running the `build.rs`, so the available token could be `0` when we obtained the jobserver via `jobserver::Client::from_env` and causing deadlock.

To prevent this, I added this piece of code which release one token into the jobserver, but it cannot be re-acquired unless I added back the global state for the obtained token.

Reply:
Oops, the CI is still in deadlock.